text/0000-break-lines-with.md

```diff
@@ -0,0 +1,95 @@
+- Feature Name: break-lines-with
+- Start Date: 2017-08-09
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+This RFC allows the users to unify linebreaks to be used in a string through method `break_lines_with()`.
+
+# Motivation
+[motivation]: #motivation
+
+Programs sometimes need to know about their working environment to do their job properly. Linebreak convention is one of the platform-related issues: It differs for Windows (CR LF) and \*NIX (LF). Such discrepancy can lead to problems easily, especially when a program needs to communicate with aged third-party libraries.
+
+As a system programming language, it would be good for Rust to know the system it works on.
+
+# Detailed design
+[design]: #detailed-design
+
+This RFC mainly introduces method `break_lines_with()` to primitive `str`. The method replaces all recognizable linebreaks with new linebreak specified, and it will not consume any linebreaks. Here is a simplified implementation of it:
```

Comment:
What advantage does this have over plain old `replace`? I don't see any.

Also, most practical code never wants to have the complete text in a single string anyway. Usually one either pulls lines by one from a `BufRead` or push them by one to a `BufWrite`.

Reply:
I would say performance, especially when the incoming string's libebreaks are unknown. e.g. I want all `LF` in my string to be replaced by `CRLF` but the method accepts not only `LF`-broken strings. In this case I can't simply replace `LF` with `CRLF`, since there is an `LF` in `CRLF` and the result will be like `CRCRLF`. Then I have to replace all `CRLF` to `LF` temporary beforehand.

Reply:
I've been thinking about your suggestion also. I've tried to make a trait `WriteLn` that allows users to specify linebreaks and write a line without using `writeln!`, as it's linebreak is always '\n'.

But my opinion is, Rust didn't explicitly distinguish text and binary reading/writing, since `OpenOption` didn't have a method like `text()` and `Read`, `Write` have both binary and text reading/writing methods. It's not `Read` or `Write`'s job to break lines. That's why I added the method to `str`.

I will update the RFC when I come up with something better.

Reply:
@PENGUINLIONG There is [`BufRead::read_line`](https://doc.rust-lang.org/std/io/trait.BufRead.html#method.read_line).

I don't see any point introducing `WriteLn` since you could just write

```rust
write!(output, "blah{newline}", newline="\r\n");
```

Reply:
@kennytm Sorry for missing that out.

But I think, it somehow reflected the need for `WriteLn` and `ReadLn`: We have a way to input per-line and we are seeking for a way to output per-line conveniently. This abstraction of per-line functionalities might be able to solve the problem.